### ISSUE TITLE
Minions get shuffled pillar data

### ISSUE THREAD

**Issue by ninja-** [saltstack/salt#39359]
### Description of Issue/Question

Minions can sometimes get shuffled pillar data.
It results in config files being changed for no reason.

It's not always shuffled and this depends on some factor I don't yet know.

```
----------
          ID: php_fpm_conf_config
    Function: file.managed
        Name: /etc/php/7.0/fpm/php-fpm.conf
      Result: True
     Comment: File /etc/php/7.0/fpm/php-fpm.conf updated
     Started: 15:27:39.418104
    Duration: 99.62 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -6,6 +6,6 @@
                   
                       
                   [global]
                  +    error_log = /var/log/php7.0-fpm.log
                       pid = /var/run/php7.0-fpm.pid
                  -    error_log = /var/log/php7.0-fpm.log
                   include = /etc/php/7.0/fpm/pool.d/*.conf
----------
          ID: ssh_config
    Function: file.managed
        Name: /etc/ssh/ssh_config
      Result: True
     Comment: File /etc/ssh/ssh_config updated
     Started: 15:27:37.381686
    Duration: 127.292 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -2,10 +2,10 @@
                   # It will be overwritten by salt!
                   
                   Host *
                  -    HashKnownHosts yes
                  +    GSSAPIAuthentication yes
                       SendEnv LANG LC_*
                       GSSAPIDelegateCredentials no
                  -    GSSAPIAuthentication yes
                  +    HashKnownHosts yes

```

etc. it's not related to a single formula
### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)
Not sure yet where does the randomness come from.
It might be possible that it's triggered after master restart.

Simplest formula that shows the problem when triggered
```
----------
          ID: rsync_config
    Function: file.managed
        Name: /etc/_rsyncd.conf
      Result: True
     Comment: File /etc/_rsyncd.conf updated
     Started: 15:27:37.073386
    Duration: 97.713 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -1,22 +1,22 @@
                   
                  +timeout = 300
                   max connections = 1000
                  -timeout = 300
                  +
                  +[salt]
                  +
                  +gid = root
                  +uid = root
                  +comment = Private Salt deploy data
                  +list = True
                  +path = /home/xxx/env/
                  +read only = True
                   
                   [xxx_tmp]
                   
                  +gid = root
                   uid = root
                  +comment = migration data
                   list = True
                  -gid = root
                  -comment = migration data
                  +path = /home/xxx/tmp/
                   read only = False
                  -path = /home/xxx/tmp/
                   
                  -[salt]
                  -
                  -uid = root
                  -list = True
                  -gid = root
                  -comment = Private Salt deploy data
                  -read only = True
                  -path = /home/xxx/env/
                  -
```
rsync_server.sls
```
rsync_config:
  file.managed:
    - name: /etc/_rsyncd.conf
    - template: jinja
    - source: salt://rsync/rsyncd.conf.jinja
    - require:
      - pkg: rsync
```

rsyncd.conf.jinja
```
{% set rsync = pillar.get('rsync',{}) -%}
{%- for key, value in rsync.server.conf.items() -%}
{{ key }} = {{ value }}
{% endfor -%}

{%- for name, share in rsync.server.shares.items() %}
[{{ name }}]
{% for key, value in share.items() %}
{{ key }} = {{ value }}
{%- endfor %}
{% endfor %}
```

Basically, the bad data is returned from pillar.get...or is the problem caused by items() call?
The pillar is supposed to be an OrderedDict - is it still the case?

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
Reproducable on both develop and 2016.11.1, python3

**Comment by ninja-** [comment]
PyYAML by default loads things unordered into a default dict, and I see nothing in salt's code which could change that behaviour when loading yaml for pillar/states...

Ref:
http://stackoverflow.com/questions/31605131/dumping-a-dictionary-to-a-yaml-file-while-preserving-order
http://stackoverflow.com/questions/5121931/in-python-how-can-you-load-yaml-mappings-as-ordereddicts
https://github.com/saltstack/salt/blob/develop/salt/renderers/yaml.py

**Comment by ninja-** [comment]
https://github.com/saltstack/salt/blob/develop/salt/utils/yamlloader.py#L38

dictclass = dict
could replacing dict with OrderedDict solve this problem?

**Comment by gtmanfred** [comment]
What do your pillar files look like?  are they only yaml, or do they load from a map.jinja file or something that could have a dictionary in it?

Thanks,
Daniel

**Comment by ninja-** [comment]
The pillar files are pure yaml

**Comment by peter-slovak** [comment]
Perhaps sorting the resulting pillar items would help:
http://jinja.pocoo.org/docs/2.9/templates/#dictsort

So an example change in the first `for` loop would be:
`{%- for item in rsync.server.conf|dictsort -%}`
and then access key and value positionally, since dictsort yields (key, value) pairs according to the docs. Or have a look at `sort`, too: http://jinja.pocoo.org/docs/2.9/templates/#sort

**Comment by ninja-** [comment]
@peter-slovak not sure about dictsort but sort is broken with python3 on OrderedDict

**Comment by pcgod** [comment]
This is still an issue with minion running on Python 3.
I did some digging into why that's happening and there are multiple reasons: YAML pillars are read as OrderedDict on the master, sent as msgpack and deserialized into dicts which do change often in Python 3. The byte/unicode conversion for Py3 also builds another dict with a different order. Next problem is that a lot of states use import_yaml in jinja (i.e. openssh formula) to get default values, which are read into a dict and merging those with the pillar data from the master creates a similiar problem even after the transfer issue is fixed.

**Comment by pcgod** [comment]
https://github.com/pcgod/salt/commit/97e025a57344f1a44d86d5faf9fd3b8b4256be63 fixes the problem for some states. It doesn't fix those using `import_yaml` or any other import in e.g. file.managed templates.
